File: cpp/snippet-training/ListRemoveDuplication.cpp
// Created by Paul Gonzalez Becerra

#include<iostream>

using namespace std;

// Variables
int	size;
char **items;

// Methods

// Gets the index of the given value
int getIndexOf(char * value)
{
	for(int i= 0; i< size; i++)
	{
		if(value== items[i])
			return i;
	}

	return -1;
}

// Removes all the duplicates of the given value
void removeDuplicatesOf(char * value, int index)
{
	for(int i= size-1; i>= 0; i--)
	{
		if(i== index)
			continue;
		if(value== items[i])
			items[i]=	(char*)0;
	}
}
void removeDuplicatesOf(char * value)	{	removeDuplicatesOf(value, getIndexOf(value));	}

// Removes any and all duplicates
void removeAllDuplicates()
{
	// Variables
	int	i=	0;

	while(i< size)
	{
		if(items[i]!= (char*)0)
			removeDuplicatesOf(items[i], i);
		i++;
	}
}

// Starts the app
void main(int argc, char * args[])
{
	size=	argc;
	items=	args;

	cout<< "Original array: ";
	for(int i= 0; i< size; i++)
	{
		if(i== size-1)
			cout<< items[i]<< endl;
		else
			cout<< items[i]<< ", ";
	}
	cout<< "Non-duplicate array: ";
	removeAllDuplicates();
	for(int i= 0; i< size; i++)
	{
		if(i== size-1)
			cout<< items[i]<< endl;
		else
			cout<< items[i]<< ", ";
	}
}

// End of File

File: cs/ListRemoveDuplication.cs
// Created by Paul Gonzalez Becerra

using System;

public static class Start
{
	// Static Methods
	
	// Starts the app
	public static void Main(string[] args)
	{
		// Variables
		ListRemoveDuplication	lrd;
		
		if(args.Length== 0)
		{
			Console.WriteLine("Please enter an array of integers in the command prompt");
			return;
		}
		lrd=	new ListRemoveDuplication(args);
		
		Console.Write("Original array: ");
		for(int i= 0; i< lrd.size; i++)
		{
			if(i== lrd.size-1)
				Console.WriteLine(lrd.items[i]);
			else
				Console.Write(lrd.items[i]+", ");
		}
		Console.Write("Non-duplication array: ");
		lrd.removeAllDuplicates();
		for(int i= 0; i< lrd.size; i++)
		{
			if(i== lrd.size-1)
				Console.WriteLine(lrd.items[i]);
			else
				Console.Write(lrd.items[i]+", ");
		}
	}
}

public class ListRemoveDuplication
{
	// Variables
	public string[]	items;
	public int	size;
	
	// Constructors
	
	public ListRemoveDuplication(string[] array)
	{
		items=	array;
		size=	array.Length;
	}
	
	// Methods
	
	// Removes all the duplicates of the given value
	public void removeDuplicatesOf(string value, int index)
	{
		for(int i= size-1; i>= 0; i--)
		{
			if(i== index)
				continue;
			if(value== items[i])
				items[i]=	null;
		}
	}
	public void removeDuplicatesOf(string value)	{	removeDuplicatesOf(value, getIndexOf(value));	}
	
	// Gets the index of the given value
	public int getIndexOf(string value)
	{
		for(int i= 0; i< size; i++)
		{
			if(value== items[i])
				return i;
		}
		
		return -1;
	}
	
	// Removes any and all duplicates within the array stored
	public void removeAllDuplicates()
	{
		// Variables
		int	i=	0;
		
		while(i< size)
		{
			if(items[i]!= null)
				removeDuplicatesOf(items[i], i); // remove(i);
			i++;
		}
	}
}

// End of File